### ISSUE TITLE
Warning mismatch?  rule _ is just a `seq` or `choice` rule with a single element.

### ISSUE THREAD

**Issue by sogaiu** [tree-sitter/tree-sitter#3455]
### Problem

Since 252e2a4bc09c14ef5289fad89acd2a92f02aa66a, `tree-sitter generate --no-bindings` produces the following warning for [tree-sitter-clojure](https://github.com/sogaiu/tree-sitter-clojure/commit/f4236d4da8aa92bc105d9c118746474c608e6af7):

```
Warning: rule char_lit is just a `seq` or `choice` rule with a single element. This is unnecessary.
```

See [this comment](https://github.com/tree-sitter/tree-sitter/pull/2577#discussion_r1652328901) for related discussion.

AFAIU, the warning text does not apply in tree-sitter-clojure's case, as [`char_lit`'s definition](https://github.com/sogaiu/tree-sitter-clojure/blob/f4236d4da8aa92bc105d9c118746474c608e6af7/grammar.js#L338-L339) is:

```js
    char_lit: $ =>
    CHARACTER,
```

and [`CHARACTER` is defined earlier](https://github.com/sogaiu/tree-sitter-clojure/blob/f4236d4da8aa92bc105d9c118746474c608e6af7/grammar.js#L189-L194) as:

```js
const CHARACTER =
      token(seq("\\",
                choice(OCTAL_CHAR,
                       NAMED_CHAR,
                       UNICODE,
                       ANY_CHAR)));
```

Here the `seq` takes two arguments and the `choice` takes four.  Neither take a single argument.

### Steps to reproduce

0. Ensure appropriate version of tree-sitter cli is installed.  Likely commits from (and including) 252e2a4bc through 3095fbe0 (most recent at time of writing) will work.

1. Clone and checkout appropriate tree-sitter-clojure

    ```
    git clone https://github.com/sogaiu/tree-sitter-clojure
    cd tree-sitter-clojure
    git checkout f4236d4da8aa92bc105d9c118746474c608e6af7
    ```

2. Use the `generate` subcommand

    ```
    tree-sitter generate --no-bindings
    ```

3. Observe warning:

    ```
    Warning: rule char_lit is just a `seq` or `choice` rule with a single element. This is unnecessary.
    ```

### Expected behavior

No warning or if there is something incorrect, a different warning message.

### Tree-sitter version (tree-sitter --version)

tree-sitter 0.22.6 (3095fbe07bbcc605910b7ad8bbd6313e0a07e678)

### Operating system/version

Void Linux

**Comment by clason** [comment]
I think a similar false warning happens with https://github.com/hankthetank27/tree-sitter-liquid, which uses a function to construct some nodes, e.g.:

https://github.com/hankthetank27/tree-sitter-liquid/blob/0419ac4868585320eee8615c90b864a1b04ef433/grammar.js#L357

**Comment by sogaiu** [comment]
Some other possibly relevant links for liquid:

* https://github.com/hankthetank27/tree-sitter-liquid/blob/0419ac4868585320eee8615c90b864a1b04ef433/grammar.js#L495-L503
* https://github.com/hankthetank27/tree-sitter-liquid/blob/0419ac4868585320eee8615c90b864a1b04ef433/grammar.js#L515
* https://github.com/hankthetank27/tree-sitter-liquid/blob/0419ac4868585320eee8615c90b864a1b04ef433/grammar.js#L617-L630

The last one shows:

```js
    _tablerow: seq(
      rules.wrapper(
        'tablerow',
        field('item', $.identifier),
        'in',
        $._iterator,
      ),

      field('body', alias(repeat(rules.node), $.block)),

      prec.right(
        rules.wrapper('endtablerow'),
      ),
    ),
```

Looks like a `seq` with 3 arguments.

**Comment by amaanq** [comment]
It's not a false positive, but it is hard to debug and find imo... If you expand the constants you've defined, you'll see there *is* a single seq right here - https://github.com/sogaiu/tree-sitter-clojure/blob/f4236d4da8aa92bc105d9c118746474c608e6af7/grammar.js#L161

I will try to work on improving the error message (e.g. adding context of what is inside the single seq/seeing if i can expand it and pinpoint to the user where it is)

**Comment by clason** [comment]
Yeah, found the [hidden spot in tree-sitter-liquid](https://github.com/hankthetank27/tree-sitter-liquid/pull/2/files#diff-919ac210accac9ecc55a76d10a7590e3d85ca3f0e165b52d30f08faee486d0cb#503) as well. So the check is fine, but more verbose error messages would be welcome. In particular, the wording should make clear that the generated parser is different from the one expected.

**Comment by amaanq** [comment]
@sogaiu @clason what are your thoughts on this? 

![image](https://github.com/tree-sitter/tree-sitter/assets/29718261/263b0689-4cf9-4671-83e1-89ff530f8b60)

(note the second one I intentionally added a seq wrapping `~@` to show what it looks like where named symbols are used. This isn't also 100% accurate, since it's reconstructing the resolved representation into the JavaScript form, but it should look almost identical to the source code.

**Comment by sogaiu** [comment]
Thanks for finding the culprit.

The choice of "contains" seems better to me than "is just".

Do you think it's too wordy to say something like: "contains or may expand to a seq or choice rule with a single element"?

I think "may expand to" captures some of the spelunking we all seem to be doing and perhaps it could serve as a hint to future readers of these warnings (^^;

**Comment by clason** [comment]
> (note the second one I intentionally added a seq wrapping `~@` to show what it looks like where named symbols are used.) Also, this isn't 100% accurate, since it's reconstructing the resolved representation into the JavaScript form, but it should look almost identical to the source code.

I think if you can't make it accurate, it's not worth the effort. I would suggest simply rewording the warning to

"rule `foo` contains ..."

(quote the rule to make it easier to see; use "contains" rather than "is" to indicate that it could be in a subrule). 

If you back out the behavior change (which I really, _really_, recommend), maybe replace "Warning" with "Note" and replace "This is unnecessary" with "Consider removing the wrapper to simplify the grammar."

**Comment by clason** [comment]
Since the behavior change was reverted for 0.23.0, I think this can be closed now.
